### PR TITLE
Shim 'libc::getrandom' in addition to 'libc::syscall(libc::SYS_getrandom)'

### PR DESCRIPTION
**Pull request by Aaron1011**

### CODE REVIEW

File: src/shims/foreign_items.rs [hunk 1]
```diff
@@ -969,3 +967,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         return Ok(None);
     }
 }
+
+// Shims the linux 'getrandom()' syscall
+fn linux_getrandom<'mir, 'tcx>(this: &mut MiriEvalContext<'mir, 'tcx>,
+                               args: &[OpTy<'tcx, Tag>],
+                               dest: PlaceTy<'tcx, Tag>) -> InterpResult<'tcx> {
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]: Please follow usual Rust formatting:
```rust
fn foo(
  arg1: type1,
  arg2: type2,
) -> ret_type {
    body
}
```

File: src/shims/foreign_items.rs [hunk 2]
```diff
@@ -969,3 +967,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         return Ok(None);
     }
 }
+
+// Shims the linux 'getrandom()' syscall
+fn linux_getrandom<'mir, 'tcx>(this: &mut MiriEvalContext<'mir, 'tcx>,
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]:
```suggestion
fn linux_getrandom<'tcx>(this: &mut MiriEvalContext<'_, 'tcx>,
```

File: src/shims/foreign_items.rs [hunk 3]
```diff
@@ -969,3 +967,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+                               args: &[OpTy<'tcx, Tag>],
+                               dest: PlaceTy<'tcx, Tag>) -> InterpResult<'tcx> {
+    let ptr = this.read_scalar(args[0])?.not_undef()?;
+    let len = this.read_scalar(args[1])?.to_usize(this)?;
+
+
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]: Please keep code "paragraph" separation to one empty line

**Comment by RalfJung** [comment]
Would also be nice to have a test for this. You can use [this old compile-fail test](https://github.com/rust-lang/miri/blob/a6f9bbc9e22be612849508552271e98bb0ae62bf/tests/compile-fail/getrandom.rs) as a basis for a run-pass test (though `// only-linux` makes more sense than the two `// ignore-` headers).

This should ideally test the syscall and the function, and also test the NULL behavior fixed by the other PR (we can just wait until that one landed).

**Comment by RalfJung** [comment]
As expected, [Linux CI fails](https://travis-ci.com/rust-lang/miri/jobs/222298485). This is good news, it means the test is run. :D

Could you add a call to `libc::getrandom` to the test?

File: tests/run-pass/linux-getrandom.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// only-linux: Uses Linux-only APIs
+
+#![feature(rustc_private)]
+extern crate libc;
+
+fn main() {
+    let mut buf = [0u8; 5];
+    unsafe {
+		assert_eq!(libc::syscall(libc::SYS_getrandom, 0 as *mut libc::c_void, 0 as libc::size_t, 0 as libc::c_uint), 0);
```
Comment (RalfJung) [inline tests/run-pass/linux-getrandom.rs]: Something seems to be wrong with indentation of this line?

**Comment by Aaron1011** [comment]
@RalfJung: It looks like we're trying to run that test on OS X, despite the `only-linux` annotation

**Comment by RalfJung** [comment]
That's weird...

**Comment by RalfJung** [comment]
Looks like compiletest_rs does not implement `only-`. :/

So, we'll have to go with the two `ignore-` then. Please add a comment explaining why we don't use `only`.

File: tests/run-pass/linux-getrandom.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// Unfortunately, compiletest_rs does not support 'only-linux',
+// so we need to ignore Windows and OS X instead
```
Comment (RalfJung) [inline tests/run-pass/linux-getrandom.rs]:
```suggestion
// so we need to ignore Windows and macOS instead.
```
It's a sentence, it should end in a full stop. (Also applies to some other comments in this PR.)

We have historically not been consistent about this in Miri, but we are trying to be consistent for new PRs.

File: src/shims/foreign_items.rs [hunk 6]
```diff
@@ -293,22 +293,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-                        let len = this.read_scalar(args[2])?.to_usize(this)?;
-
-                        // The only supported flags are GRND_RANDOM and GRND_NONBLOCK,
-                        // neither of which have any effect on our current PRNG
-                        let _flags = this.read_scalar(args[3])?.to_i32()?;
-
-                        this.gen_random(ptr, len as usize)?;
-                        this.write_scalar(Scalar::from_uint(len, dest.layout.size), dest)?;
+                        // The first argument is the syscall id,
+                        // so skip over it
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]:
```suggestion
                        // so skip over it.
```

File: src/shims/foreign_items.rs [hunk 7]
```diff
@@ -969,3 +967,21 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         return Ok(None);
     }
 }
+
+// Shims the linux 'getrandom()' syscall
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]:
```suggestion
// Shims the linux 'getrandom()' syscall.
```

File: src/shims/foreign_items.rs [hunk 8]
```diff
@@ -969,3 +967,21 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+fn linux_getrandom<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    let ptr = this.read_scalar(args[0])?.not_undef()?;
+    let len = this.read_scalar(args[1])?.to_usize(this)?;
+
+    // The only supported flags are GRND_RANDOM and GRND_NONBLOCK,
+    // neither of which have any effect on our current PRNG
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]:
```suggestion
    // neither of which have any effect on our current PRNG.
```

**Comment by RalfJung** [comment]
Thanks! And CI also looks green. So let's

@bors r+
